libbpf-tools/hardirqs.bpf.c

```diff
@@ -0,0 +1,107 @@
+// SPDX-License-Identifier: GPL-2.0
+// Copyright (c) 2020 Wenbo Zhang
+#include "vmlinux.h"
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_tracing.h>
+#include "hardirqs.h"
+#include "bits.bpf.h"
+#include "maps.bpf.h"
+
+#define MAX_ENTRIES	256
+
+const volatile bool targ_dist = false;
+const volatile bool targ_ns = false;
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, struct irq_key);
+	__type(value, u64);
+	__uint(map_flags, BPF_F_NO_PREALLOC);
+} irqs SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERCPU_ARRAY);
+	__uint(max_entries, 1);
+	__type(key, u32);
+	__type(value, u64);
+} start SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, struct irq_key);
+	__type(value, struct hist);
+	__uint(map_flags, BPF_F_NO_PREALLOC);
+} hists SEC(".maps");
+
+static struct hist zero_hist;
+
+SEC("tracepoint/irq/irq_handler_entry")
+int handle__irq_handler(struct trace_event_raw_irq_handler_entry *ctx)
+{
+	struct irq_key key = {};
+	u64 *count, zero = 0;
+
+	bpf_probe_read_kernel_str(&key.name, sizeof(key.name), ctx->__data);
```

Comment:
why using symbolic name as a key? It's inconvenient and slow. There seems to be `irq` int field, use that for key look up?

Reply:
> why using symbolic name as a key

Because I think irq cannot distinguish shared interrupts，such as (cp fromFrom http://static.lwn.net/images/pdf/LDD3/ch10.pdf ):

> All the handlers installed for the same interrupt number appear on the same line of
> /proc/interrupts. The following output (from an x86_64 system) shows how shared
> interrupt handlers are displayed:
>  CPU0
>  0: 892335412 XT-PIC timer
>  1: 453971 XT-PIC i8042
>  2: 0 XT-PIC cascade
>  5: 0 XT-PIC libata, ehci_hcd
>  8: 0 XT-PIC rtc
>  9: 0 XT-PIC acpi
>  10: 11365067 XT-PIC ide2, uhci_hcd, uhci_hcd, SysKonnect SK-98xx, EMU10K1
>  11: 4391962 XT-PIC uhci_hcd, uhci_hcd
>  12: 224 XT-PIC i8042
>  14: 2787721 XT-PIC ide0
>  15: 203048 XT-PIC ide1
> NMI: 41234
> LOC: 892193503
> ERR: 102
> MIS: 0
> This system has several shared interrupt lines. IRQ 5 is used for the serial ATA and
> IEEE 1394 controllers; IRQ 10 has several devices, including an IDE controller, two
> USB controllers, an Ethernet interface, and a sound card; and IRQ 11 also is used by
> two USB controllers.

Reply:
oh, ok, that makes sense, thanks!